Fix bench-e2e single mode and keep results

ch1bo opened the pull request:
This fixes two issues with the `bench-e2e` binary / benchmark:

* Running in `single` mode was not working because of a `FeeTooSmallUTxO` error
* The `results.csv` is written into a temporary directory and removed, which makes plotting impossible.

I was in the mood of some refactoring so this contains also various other changes I encountered while working on the code and I was [tidying up](https://tidyfirst.substack.com/p/the-life-changing-magic-of-tidying) a bit.

The refactoring separated hydra node and payment keys further, which requires the datasets to be re-generated. I took the freedom to generate with `--scaling-factor 10` which results in `300` transactions per client. Should be long enough to identify regressions, with **hopefully 10x shorter benchmark time** in CI.

Another benefit of this separation is that it naturally led to reducing the assumptions of the `demo` mode by not seeding the hydra node cardano keys, but re-using `seed-devnet.sh` and consequently looser coupling between the workload and container setup in our network test workflow.

I'm not 100% happy with how the bench is now requiring the `--output-directory` to be empty, and in turn the whole state will be captured as an artifact of our CI. Instead, making the state directory always a /tmp path and retained in case of errors (or configurable with `--state-directory`) would be better. But that can go into another PR .. another time.

---

* [x] CHANGELOG updated
* [x] Documentation updatedx (README)
* [x] Haddocks updated
* [x] No new TODOs introduced or explained herafter
  - Two XXX notes of what to improve further

ch1bo committed:
Remove redundant bench-e2e mode of single + workdir set

This is redundant and can be achieved by using the 'datasets'
subcommand.

ch1bo committed:
Write generated dataset to outputDirectory if given

Before it was written to a random temporary directory, which makes it
annoying to generate datasets with this mode.

ch1bo committed:
Small module re-org in HydraNode.hs

ch1bo committed:
Drop need of party in benchmark scenario

ch1bo committed:
Use a bit more non empty lists in hydra-cluster

ch1bo committed:
Switch to only use self-transfers in benchmarks

This reduces some code duplication without much loss of
expressiveness (which key we use does not matter).

ch1bo committed:
Separate client keys and hydra node keys

ch1bo committed:
Fix hydra-cluster bench single by setting a hard-coded fee

This is not ideal, but a lot simpler than doing proper fee calculation.
It's unclear why fee calculation was removed before, it is needed when
running benchmark scenarios.

ch1bo committed:
Write results.csv in --output-directory

They hydra-cluster benchmarks now only uses a single directory to store
the whole state, which is temporary unless a specific output-directory
is requested.

ch1bo committed:
Fail if output directory is not empty

ch1bo committed:
Update datasets with new format

Same transaction style (single repending txs), but deliberately smaller
length of transactions (3000 -> 300) to have shorter benchmark
run-times, while sequence should be long enough to identify regressions.

Generated with invocations:

cabal run bench-e2e -- single --cluster-size 1 --scaling-factor 10

and

cabal run bench-e2e -- single --cluster-size 3 --scaling-factor 10

Plus some manual amending of the JSON to contain a "title".

noonio commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -145,40 +129,32 @@ benchDemo networkId nodeSocket timeoutSeconds hydraClients workDir dataset@Datas
-  Set Party ->
-  HydraClient ->
-  [HydraClient] ->
+  NonEmpty HydraClient ->
```
Could use a comment at least to say that the head is the leader always.

noonio commented on hydra-cluster/hydra-cluster.cabal:
```diff
@@ -1,6 +1,6 @@
 cabal-version:      3.0
 name:               hydra-cluster
-version:            0.19.0
```
why bump this now? is it just always equal to the "hydra node" version? if so, should just leave it until we release?

ch1bo replied:
This is a practice I occasionally follow to indicate how much a package changed, but yes at release time we would bump all of them to the same version (typically a `0.x` bump as it is breaking / adds features). So I usually update package versions ahead of time alongside keeping the changelog updated during development.

As this PR is dropping the `--work-directory` part of `single` it is a breaking changed and would result in semantically versioned `0.20.0` here.

noonio reviewed (approved):
Minor comments; happy to merge if all the tests pass!

noonio commented:
In fact the network tests are failing - https://github.com/cardano-scaling/hydra/actions/runs/11241196905/job/31255255123?pr=1693

ch1bo committed:
Start with an empty output directory

noonio committed:
Only show quantiles if they can be computed

ch1bo committed:
Use and seed hydraNodeKeys in demo mode

As before, the bench-e2e does not assume the hydra node keys to be
seeded. This ties the way bench-e2e binary (which hard-codes Alice, Bob
and Carol) to the configurable list of --hydra-client to connect to.

ch1bo committed:
Use seed-devnet and not assume node keys in bench-e2e

This decouples the bench-e2e binary which just produces load and
provides statistics more from how the hydra-nodes are run.

Now the only assumption is that the
'hydra-cluster/config/credentials/faucet.sk' owns funds on the given
network.

ch1bo committed:
Add a comment to the benchmark scenario
